### ISSUE TITLE
Batch  normalisation with shared vision model sometimes causes NaN.

### ISSUE THREAD

**Issue by karolbadowski** [tensorflow/tensorflow#43603]
Hello.

**Intorduction:**

I am writing a neural network that analyses video. For analyzing different frames I use shared vision model (the same set of layers and weights is applied over various images) and then I concatenate these columns into one model.

Why do I use Shared Vision Model?
As 3D convolutions and TimeDistributed models are not supported in conversion to TFLite that goes in pair with Tensorflow 1...
I need to resort to using a **Shared Vision Model**.

**When Problem does NOT occur:**
Training goes without any problems when I do not use Batch Normalization.
Training goes without problems if I use Batch Normalization in layers of neural network that are after the consatenation of frames.
Training also goes well if I train columns of neural network that analyze single images, with batch normalization, using TimeDistributed Embedding.
Training goes well if I use for the same purpose, 3D convolutions with depth (along time axis) set to 1.

**When Problem occurs:**
If I train neural network with usage of Batch Normalization on layers that belong to Shared Vision Model... training sometimes ends with NaN loss.

When in the same situation I switch batch normalization off, the problem never occurs.

Here is an example of training with shared vision model and BatchNormalization layers activated:
![obraz](https://user-images.githubusercontent.com/4332432/94397173-8fbf4880-0163-11eb-8bb5-8c24ef77b004.png)

My questions are:

- What could have gone wrong?

- Is it possible that batch normalization with usage of Shared Vision model has problem with accumulation of values from various instances?
- Is it possible that Batch Normalization returns NaN when all samples in a batch returned the same value (so variance is 0 and difference between mean and max is 0)?

Could NaN have came from a bug in implementation of Keras BatchNormalization() layer?
(I have seen a suspicious line of code:) 
`# sample variance - unbiased estimator of population variance`
`variance *= sample_size / (sample_size - (1.0 + self.epsilon))`

For me it seems that internal brackets cause that epsilon is subtracted instead of added, which for sample_size 1 will cause variance to be devided by a negative value.
Inside of K.normalize_batch_in_training() -> _broadcast_normalize_batch_in_training() -> tf.nn.batch_normalization(), there is line of code:
inv = math_ops.rsqrt(variance + variance_epsilon)

I guess that in some scenarios, negative value of variance could lead to rsqrt() from a negative value, which results in NaN...

NaN appearance during training is so unpredictable, dependent on outputs of previous layers and on combination of random batch, that I am not able to exactly provide inputs that cause mentioned problem.
However analysis what could have gone wrong (division by zero, root from minus value, or something else) could lead to finding a volatile spot.

I do not knwo if the one I have fund is the one...
I do nto know why this occurs for me only randomly and only in case of usage of Shard Vision Model.

Could it be wrong aggregation of data, that occurs with Shared Vision Model, but not with TimeDistributed wrapper and not with 3DConvolutions which have more stable data aggregation?

Could this error occur when all activations of layer predecessing to BatchNormalization() are the same across the batch?

That issue is important for me.

**Comment by karolbadowski** [comment]
I see that there are people who have experienced the same issue for past 2 years.
https://github.com/keras-team/keras/issues/11927

Therefore I am not the only one - they have posted samples related maybe not to video analysis, but to Triplet and Siamese networks, which are the most common use cases of Shared Vision Model.

I strongly suspect there is something wrong with moving mean not being aggregated well across the samples.
Real solution needed urgently.

**Comment by bhack** [comment]
Can you try to debug your case with https://www.tensorflow.org/tensorboard/debugger_v2?hl=en#using_debugger_v2_gui_to_find_the_root_cause_of_nans ?

**Comment by karolbadowski** [comment]
![obraz](https://user-images.githubusercontent.com/4332432/94397173-8fbf4880-0163-11eb-8bb5-8c24ef77b004.png)
I will try to find it - luckily, I had TensorBoard turned on during one of such occurrences.

**Comment by karolbadowski** [comment]
Hello @bhack .

Maybe it will be the easiest to reproduce the issue with simple code samples of siamese network provided in thread keras-team/keras#11927 , as according to thread participants problem still occurs in newest TensorFlow and their code sample reveals NaN quickly.


I will try my best to activate debugger to help finding the source with my older version of tensorflow (as issue seems so still be the same).

I have a following problem whne I activate debugger under Kears 2.2.4 with tensorflow-gpu 1.12.3 :

First I activate my tensorboard:
`tensorboard --logdir ./logs --debugger_port 999`

My TensorBoard advised me to use the following code:
```
import tensorflow as tf
from tensorflow.python import debug as tf_debug
import keras

keras.backend.set_session(
    tf_debug.TensorBoardDebugWrapperSession(tf.Session(), "karol-Precision-7520:999"))

...........(and here my training pipeline, creating the Keras model and running the training).......................
```
I pasted it in the beginning of the code.

When I tarted my training pipeline, during model building, the code in that debug mode stopped with the following call stack (surprisingly, on the BatchNormalization() layer in the call stack):

>   File "/home/karol/Repozytoria/xxx/common/models/merged_2_stages_saveable_splitable.py", line 94, in _acquire_images_pair_column
>     if batch_normalisation: images_pair_column = K.layers.BatchNormalization()(images_pair_column)
>   File "/home/karol/.local/lib/python3.6/site-packages/keras/engine/base_layer.py", line 457, in __call__
>     output = self.call(inputs, **kwargs)
>   File "/home/karol/.local/lib/python3.6/site-packages/keras/layers/normalization.py", line 185, in call
>     epsilon=self.epsilon)
>   File "/home/karol/.local/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 1858, in normalize_batch_in_training
>     if not _has_nchw_support() and list(reduction_axes) == [0, 2, 3]:
>   File "/home/karol/.local/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 292, in _has_nchw_support
>     gpus_available = len(_get_available_gpus()) > 0
>   File "/home/karol/.local/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 278, in _get_available_gpus
>     _LOCAL_DEVICES = get_session().list_devices()
>   File "/home/karol/.local/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 199, in get_session
>     [tf.is_variable_initialized(v) for v in candidate_vars])
>   File "/home/karol/.local/lib/python3.6/site-packages/tensorflow/python/debug/wrappers/grpc_wrapper.py", line 223, in run
>     self._sent_graph_version)
>   File "/home/karol/.local/lib/python3.6/site-packages/tensorflow/python/debug/wrappers/grpc_wrapper.py", line 63, in publish_traceback
>     send_source=True)
>   File "/home/karol/.local/lib/python3.6/site-packages/tensorflow/python/debug/lib/source_remote.py", line 209, in send_graph_tracebacks
>     graph=graph, send_source=send_source)
>   File "/home/karol/.local/lib/python3.6/site-packages/tensorflow/python/debug/lib/source_remote.py", line 174, in _send_call_tracebacks
>     stub.SendTracebacks(call_traceback)
>   File "/home/karol/.local/lib/python3.6/site-packages/grpc/_channel.py", line 826, in __call__
>     return _end_unary_response_blocking(state, call, False, None)
>   File "/home/karol/.local/lib/python3.6/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking
>     raise _InactiveRpcError(state)
> grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
> 	status = StatusCode.UNAVAILABLE
> 	details = "failed to connect to all addresses"
> 	debug_error_string = "{"created":"@1601277231.862699005","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":3981,"referenced_errors":[{"created":"@1601277231.862696725","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":394,"grpc_status":14}]}"
> >
> 
> Process finished with exit code 1

**I am wondering what I can do to accelerate the investigation.
Can I activate the debugger in TensorFlow 1 without the quoted bug?**

I can present my part of neural network (not exactly the same, as I am bound by NDA, but idea stays the same):

I am creating a shared vision model that takes 1 image on input.
It is a bunch of layers starting with:
input_tensor = `K.layers.Input(shape=(self.img_rows,self.img_cols,self.channels))`
Then there I have bunch of Conv2D, LeakyReLU and MaxPooling2D layers.
and ending with:
```
        if batch_normalisation: image_column = K.layers.BatchNormalization()(image_column)
        self.single_image_column = K.models.Model(input_tensor,image_column,name="single_image_column")
```
Later I have a model that looks at pairs of images.
It has 2 inputs:

```
recent_input_tensor = K.layers.Input(shape=self.single_image_column.output_shape[1:])
latter_input_tensor = K.layers.Input(shape=self.single_image_column.output_shape[1:])
images_pair_column = K.layers.Concatenate(axis=1 if self.channels_first else -1)([recent_input_tensor, latter_input_tensor])
```
Again, story is the same, bunch of Conv2D, LeakyReLu and MaxPooling2D layers.
Model ends with:
```
if batch_normalisation: images_pair_column = K.layers.BatchNormalization()(images_pair_column)
self.images_pair_column = K.models.Model([recent_input_tensor,latter_input_tensor],images_pair_column,name="images_pair_column")
```
And analogically, I have a third brick of the model, that concatenates 2D feature maps information from 2 images_pair columns (my total field of view is 3 video frames).
There is again bunch of Conv2D and LeakyReLu layers, there is also a global pooling and BatchNormalization.
In the end I have a few fully connected Dense layers and 1 neuron output of regression with mse loss function.

I think that code sampels presented in keras-team/keras#11927 will be much simpler to debug, as they get the same kind of error, with a classical Siamese model (much less complex).

What is common among all of the cases:
- Problem happens when BatchNormalization() layer is in a shared model that is applied over 2 or more (in my case more) input images.
![obraz](https://user-images.githubusercontent.com/4332432/94408064-97d3b400-0174-11eb-9cc8-1866bef87b2c.png)

- Output layers from various images, processed by shared vision model are concatenated. When BatchNormalization() layer is in that concatenated part, it does not cause the problem. 

When I have BatchNormalization in the part of model that was concatenated (just before ending) - so there is just 1 instance...  it does not cause a problem.
![obraz](https://user-images.githubusercontent.com/4332432/94408231-cb164300-0174-11eb-8e96-ba4fa1df5d1a.png)

**Comment by bhack** [comment]
I don't think that `https://github.com/keras-team/keras/` ticket system is currently actively maintained at least until Keras will be standalone again (See https://github.com/keras-team/keras/releases/tag/2.4.0)

We need a very, very minimal but completely runnable standalone example (copy, paste and run) or Colab to reproduce your use case.
